File: server/src/utils/think-parser.js
/**
 * Strips thinking blocks from LLM responses.
 * Block delimiters are configurable per settings preset.
 */
export function stripThinkBlocks(text, startTag = '<think>', endTag = '</think>') {
  if (!text) return { narrative: '', thinking: '' };

  const thinkingParts = [];
  let narrative = text;

  // Escape regex special chars in tags
  const escStart = startTag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const escEnd = endTag.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Match complete think blocks
  const regex = new RegExp(`${escStart}([\\s\\S]*?)${escEnd}`, 'g');
  let match;
  while ((match = regex.exec(narrative)) !== null) {
    thinkingParts.push(match[1].trim());
  }
  narrative = narrative.replace(regex, '');

  // Handle unclosed think block (everything from start tag to end)
  const unclosedIdx = narrative.indexOf(startTag);
  if (unclosedIdx !== -1) {
    thinkingParts.push(narrative.slice(unclosedIdx + startTag.length).trim());
    narrative = narrative.slice(0, unclosedIdx);
  }

  return {
    narrative: narrative.trim(),
    thinking: thinkingParts.join('\n\n'),
  };
}

File: server-python/app/utils/think_parser.py
"""Extract and strip <think> blocks from LLM output."""
import re


def strip_think_blocks(
    text: str,
    start_tag: str = "<think>",
    end_tag: str = "</think>",
) -> tuple[str, str | None]:
    """Return (narrative, thinking). Handles closed and unclosed blocks."""
    if start_tag not in text:
        return text.strip(), None

    # Closed blocks: <think>...</think>
    pattern = re.escape(start_tag) + r"(.*?)" + re.escape(end_tag)
    thinking_parts: list[str] = []
    narrative = text

    for match in re.finditer(pattern, text, re.DOTALL):
        thinking_parts.append(match.group(1).strip())
        narrative = narrative.replace(match.group(0), "", 1)

    # Unclosed block: <think>... (rest of text)
    if start_tag in narrative:
        idx = narrative.index(start_tag)
        thinking_parts.append(narrative[idx + len(start_tag):].strip())
        narrative = narrative[:idx]

    thinking = "\n\n".join(thinking_parts) if thinking_parts else None
    return narrative.strip(), thinking
